File: stdlib/kdb/kdb.py
# test 631 : stdlib5.kdb

class toolbox: # for lite setting files
    def __init__(self):
        self.name = dict() # (int index)[str name]
        self.type = [ ] # type int[]
        self.ptr = [ ] # pointer int[]
        self.fmem = [ ] # mem float[]
        self.cmem = [ ] # mem complex[]
        self.bmem = [ ] # mem bytes[]

        self.working = [ ] # current names list

    # parse string
    def read(self, raw):
        # pass id setter nonstr str sharp end
        self.working = [ ]
        order = [ x + '\n' for x in raw.split('\n') ]

        for i in order:
            status, mem, name, var = 'pass', [ ], "", ""
            for j in i:
                if status == 'pass':
                    if j not in [' ', '\r']:
                        if j in ['\n', ';']:
                            status, mem, name, var = 'pass', [ ], "", ""
                        elif j == '=':
                            raise Exception(f"invalid key : {i}")
                        else:
                            status = 'id'
                            mem.append(j)
                elif status == 'id':
                    if j in ['\n', ';']:
                        status, mem, name, var = 'pass', [ ], "", ""
                    elif j not in [' ', '\r']:
                        if j == '=':
                            status = 'setter'
                            name = ''.join(mem)
                            mem = [ ]
                        else:
                           mem.append(j)
                elif status == 'setter':
                    if j not in [' ', '\r']:
                        if j in ['\n', ';']:
                            raise Exception(f"invalid value : {i}")
                        elif j == '"':
                            status = 'str'
                            mem.append(j)
                        else:
                            status = 'nonstr'
                            mem.append(j)
                elif status == 'nonstr':
                    if j not in [' ', '\r']:
                        if j == '\n':
                            var = ''.join(mem)
                            self.add(name, var, '\n')
                            status, mem, name, var = 'pass', [ ], "", ""
                        elif j == ';':
                            var = ''.join(mem)
                            self.add(name, var, ';')
                            status, mem, name, var = 'pass', [ ], "", ""
                        else:
                            mem.append(j)
                elif status == 'str':
                    if j == '\n':
                        raise Exception(f"invalid value : {i}")
                    elif j == '#':
                        status = 'sharp'
                    elif j == '"':
                        mem.append(j)
                        var = ''.join(mem)
                        status = 'end'
                    else:
                        mem.append(j)
                elif status == 'sharp':
                    if j == '#':
                        mem.append('#')
                        status = 'str'
                    elif j == 's':
                        mem.append(' ')
                        status = 'str'
                    elif j == 'n':
                        mem.append('\n')
                        status = 'str'
                    elif j == '"':
                        mem.append('"')
                        status = 'str'
                    else:
                        raise Exception(f"invalid escaping : {i}")
                elif status == 'end':
                    if j == '\n':
                        self.add(name, var, '\n')
                        status, mem, name, var = 'pass', [ ], "", ""
                    elif j == ';':
                        self.add(name, var, ';')
                        status, mem, name, var = 'pass', [ ], "", ""

    # return current DB str
    def write(self):
        out = [""] * len(self.type)
        for i in self.name:
            toadd = i
            num = self.name[i]
            tp = self.type[num]
            ptr = self.ptr[num]
            if tp // 16 == 0:
                end = '\n'
            else:
                end = '; '
            tp = tp % 16

            toadd = toadd + ' = '

            if tp == 0:
                toadd = toadd + self.conv(None)
            elif tp == 1:
                if ptr == 0:
                    toadd = toadd + self.conv(False)
                else:
                    toadd = toadd + self.conv(True)
            elif tp == 2:
                toadd = toadd + self.conv(ptr)
            elif tp == 3:
                toadd = toadd + self.conv( self.fmem[ptr] )
            elif tp == 4:
                toadd = toadd + self.conv( self.cmem[ptr] )
            elif tp == 5:
                toadd = toadd + self.conv( self.bmem[ptr] )
            elif tp == 6:
                temp = str(self.bmem[ptr], encoding='utf-8')
                toadd = toadd + self.conv(temp)
            else:
                raise Exception(f"invalid type : {tp}")
            toadd = toadd + end
            out[num] = toadd
        return ''.join(out)

    # get [index, type, ptr, value] by name
    def get(self, name):
        name = name.replace('/', '.')
        num = self.name[name]
        tp = self.type[num] % 16
        ptr = self.ptr[num]

        if tp == 0:
            var = None
            t = 'None'
        elif tp == 1:
            if ptr == 0:
                var = False
            else:
                var = True
            t = 'bool'
        elif tp == 2:
            var = ptr
            t = 'int'
        elif tp == 3:
            var = self.fmem[ptr]
            t = 'float'
        elif tp == 4:
            var = self.cmem[ptr]
            t = 'complex'
        elif tp == 5:
            var = self.bmem[ptr]
            t = 'bytes'
        elif tp == 6:
            var = str(self.bmem[ptr], encoding='utf-8')
            t = 'str'
        else:
            raise Exception(f"invalid type : {tp}")

        return [num, t, ptr, var]

    # revice data by name
    def fix(self, name, data):
        name = name.replace('/', '.')
        num = self.name[name]
        end = self.type[num] // 16
        end = end * 16

        if data == None:
            self.type[num] = end + 0
            self.ptr[num] = 0
        elif type(data) == bool:
            self.type[num] = end + 1
            if data:
                self.ptr[num] = 1
            else:
                self.ptr[num] = 0
        elif type(data) == int:
            self.type[num] = end + 2
            self.ptr[num] = data
        elif type(data) == float:
            self.type[num] = end + 3
            self.ptr[num] = len(self.fmem)
            self.fmem.append(data)
        elif type(data) == complex:
            self.type[num] = end + 4
            self.ptr[num] = len(self.cmem)
            self.cmem.append(data)
        elif type(data) == bytes:
            self.type[num] = end + 5
            self.ptr[num] = len(self.bmem)
            self.bmem.append(data)
        elif type(data) == str:
            self.type[num] = end + 6
            self.ptr[num] = len(self.bmem)
            self.bmem.append( bytes(data, encoding='utf-8') )
        else:
            raise Exception(f"invalid type : {type(data)}")

    # convert data -> kformat str
    def conv(self, data):
        if data == None:
            return 'None'
        elif type(data) == bool:
            if data:
                return 'True'
            else:
                return 'False'
        elif type(data) == int:
            return str(data)
        elif type(data) == float:
            return str(data)
        elif type(data) == complex:
            temp = str(data)
            if temp[0] == '(':
                temp = temp[1:-2] + 'j'
            else:
                temp = temp[0:-1] + 'j'
            return temp
        elif type(data) == bytes:
            temp = [0] * len(data)
            for i in range( 0, len(data) ):
                if data[i] > 15:
                    temp[i] = str( hex( data[i] ) )[2:4]
                else:
                    temp[i] = '0' + str( hex( data[i] ) )[2]
            return "'" + ''.join(temp) + "'"
        elif type(data) == str:
            data = data.replace('#', '##')
            data = data.replace(' ', '#s')
            data = data.replace('\n', '#n')
            data = data.replace('"', '#"')
            return '"' + data + '"'
        else:
            raise Exception(f"invalid type : {type(data)}")

    # add DB by name, var, end
    def add(self, name, var, end):
        name = name.replace('/', '.')
        num = 0
        while name[num] == '.':
            num = num + 1
        self.working = self.working[0:num] + [ name[num:] ]
        name = '.'.join(self.working)
        
        if end == '\n':
            tp = 0
        else:
            tp = 16

        if var == 'None':
            self.name[name] = len(self.type)
            self.type.append(tp + 0)
            self.ptr.append(0)
        elif var == 'True':
            self.name[name] = len(self.type)
            self.type.append(tp + 1)
            self.ptr.append(1)
        elif var == 'False':
            self.name[name] = len(self.type)
            self.type.append(tp + 1)
            self.ptr.append(0)
        elif var[0] == '"':
            self.name[name] = len(self.type)
            self.type.append(tp + 6)
            self.ptr.append( len(self.bmem) )
            self.bmem.append( bytes(var[1:-1], 'utf-8') )
        elif var[0] == "'":
            self.name[name] = len(self.type)
            self.type.append(tp + 5)
            self.ptr.append( len(self.bmem) )
            self.bmem.append( bytes.fromhex( var[1:-1].lower() ) )
        elif var[-1] == 'j':
            self.name[name] = len(self.type)
            self.type.append(tp + 4)
            self.ptr.append( len(self.cmem) )
            self.cmem.append( complex( var.replace('i', 'j') ) )
        elif '.' in var:
            self.name[name] = len(self.type)
            self.type.append(tp + 3)
            self.ptr.append( len(self.fmem) )
            self.fmem.append( float(var) )
        else:
            self.name[name] = len(self.type)
            self.type.append(tp + 2)
            self.ptr.append( int(var) )

    # import and set by input pylist
    def imp(self, arr):
        for i in arr:
            name = i[0]
            data = i[1]
            end = i[2]
            
            num = len(self.type)
            self.name[name] = num
            if end == '\n':
                end = 0
            else:
                end = 16
            self.type.append(0)
            self.ptr.append(0)

            if data == None:
                self.type[num] = end + 0
                self.ptr[num] = 0
            elif type(data) == bool:
                self.type[num] = end + 1
                if data:
                    self.ptr[num] = 1
                else:
                    self.ptr[num] = 0
            elif type(data) == int:
                self.type[num] = end + 2
                self.ptr[num] = data
            elif type(data) == float:
                self.type[num] = end + 3
                self.ptr[num] = len(self.fmem)
                self.fmem.append(data)
            elif type(data) == complex:
                self.type[num] = end + 4
                self.ptr[num] = len(self.cmem)
                self.cmem.append(data)
            elif type(data) == bytes:
                self.type[num] = end + 5
                self.ptr[num] = len(self.bmem)
                self.bmem.append(data)
            elif type(data) == str:
                self.type[num] = end + 6
                self.ptr[num] = len(self.bmem)
                self.bmem.append( bytes(data, encoding='utf-8') )
            else:
                raise Exception(f"invalid type : {type(data)}")

    # export to precise pylist
    def exp(self):
        out = [0] * len(self.type) # [fullname str, data, end]
        for i in self.name:
            temp = ["", "", ""]
            num = self.name[i]
            tp = self.type[num]
            ptr = self.ptr[num]

            temp[0] = i
            if tp // 16 == 0:
                temp[2] = '\n'
            else:
                temp[2] = ';'
            tp = tp % 16

            if tp == 0:
                temp[1] = None
            elif tp == 1:
                if ptr == 0:
                    temp[1] = False
                else:
                    temp[1] = True
            elif tp == 2:
                temp[1] = ptr
            elif tp == 3:
                temp[1] = self.fmem[ptr]
            elif tp == 4:
                temp[1] = self.cmem[ptr]
            elif tp == 5:
                temp[1] = self.bmem[ptr]
            elif tp == 6:
                temp[1] = str(self.bmem[ptr], encoding='utf-8')
            else:
                raise Exception(f"invalid type : {tp}")
            
            out[num] = temp
        return out

File: stdlib/kdb/kdb.go
// test632 : stdlib5.kdb

package kdb

import (
	"fmt"
	"stdlib5/kio"
	"strconv"
	"strings"
)

// pyvar, generation only
type kdbvar struct {
	Dat0 string
	Dat1 bool
	Dat2 int
	Dat3 float64
	Dat4 complex128
	Dat5 []byte
	Dat6 string
}

// kdbvar 7type setter
func Set(v interface{}) kdbvar {
	var out kdbvar
	switch x := interface{}(v).(type) {
	case nil:
		out.Dat0 = "None"
	case bool:
		out.Dat0 = "bool"
		out.Dat1 = x
	case int:
		out.Dat0 = "int"
		out.Dat2 = x
	case float64:
		out.Dat0 = "float"
		out.Dat3 = x
	case complex128:
		out.Dat0 = "complex"
		out.Dat4 = x
	case []byte:
		out.Dat0 = "bytes"
		out.Dat5 = x
	case string:
		out.Dat0 = "str"
		out.Dat6 = x
	default:
		out.Dat0 = "None"
	}
	return out
}

// for lite setting files
type toolbox struct {
	Name    map[string]int // (int index)[str name]
	Tp      []byte         // type int[]
	Ptr     []int          // pointer int[]
	Fmem    []float64      // mem float[]
	Cmem    []complex128   // mem complex[]
	Bmem    [][]byte       // mem bytes[]
	working []string       // current names list
}

// Init and return toolbox
func Initkdb() toolbox {
	var out toolbox
	out.Name = make(map[string]int)
	return out
}

// parse string
func (t *toolbox) Read(raw string) error {
	// pass id setter nonstr str sharp end
	t.working = make([]string, 0)
	order := strings.Split(raw, "\n")
	for i, r := range order {
		order[i] = r + "\n"
	}

	for _, i := range order {
		status := "pass"
		var mem []string
		name := ""
		v := ""

		for _, j := range i {
			switch status {

			case "pass":
				if (j != ' ') && (j != '\r') {
					if (j == '\n') || (j == ';') {
						status = "pass"
						mem = make([]string, 0)
						name = ""
						v = ""
					} else if j == '=' {
						return fmt.Errorf("invalid key : %s", i)
					} else {
						status = "id"
						mem = append(mem, string(j))
					}
				}

			case "id":
				if (j == '\n') || (j == ';') {
					status = "pass"
					mem = make([]string, 0)
					name = ""
					v = ""
				} else if (j != ' ') && (j != '\r') {
					if j == '=' {
						status = "setter"
						name = strings.Join(mem, "")
						mem = make([]string, 0)
					} else {
						mem = append(mem, string(j))
					}
				}

			case "setter":
				if (j != ' ') && (j != '\r') {
					if (j == '\n') || (j == ';') {
						return fmt.Errorf("invalid value : %s", i)
					} else if j == '"' {
						status = "str"
						mem = append(mem, string(j))
					} else {
						status = "nonstr"
						mem = append(mem, string(j))
					}
				}

			case "nonstr":
				if (j != ' ') && (j != '\r') {
					if j == '\n' {
						v = strings.Join(mem, "")
						t.add(name, v, "\n")
						status = "pass"
						mem = make([]string, 0)
						name = ""
						v = ""
					} else if j == ';' {
						v = strings.Join(mem, "")
						t.add(name, v, ";")
						status = "pass"
						mem = make([]string, 0)
						name = ""
						v = ""
					} else {
						mem = append(mem, string(j))
					}
				}

			case "str":
				if j == '\n' {
					return fmt.Errorf("invalid value : %s", i)
				} else if j == '#' {
					status = "sharp"
				} else if j == '"' {
					mem = append(mem, string(j))
					v = strings.Join(mem, "")
					status = "end"
				} else {
					mem = append(mem, string(j))
				}

			case "sharp":
				switch j {
				case '#':
					mem = append(mem, "#")
					status = "str"
				case 's':
					mem = append(mem, " ")
					status = "str"
				case 'n':
					mem = append(mem, "\n")
					status = "str"
				case '"':
					mem = append(mem, "\"")
					status = "str"
				default:
					return fmt.Errorf("invalid escaping : %s", i)
				}

			case "end":
				if j == '\n' {
					t.add(name, v, "\n")
					status = "pass"
					mem = make([]string, 0)
					name = ""
					v = ""
				} else if j == ';' {
					t.add(name, v, ";")
					status = "pass"
					mem = make([]string, 0)
					name = ""
					v = ""
				}
			}
		}
	}

	return nil
}

// return current DB str
func (t *toolbox) Write() (string, error) {
	out := make([]string, len(t.Tp))
	for name, num := range t.Name {
		tp := t.Tp[num]
		ptr := t.Ptr[num]
		var end string
		if tp/16 == 0 {
			end = "\n"
		} else {
			end = "; "
		}
		tp = tp % 16
		toadd := name + " = "

		var temp string
		var cnvr error
		switch tp {
		case 0:
			temp, cnvr = conv(nil)
		case 1:
			if ptr == 0 {
				temp, cnvr = conv(false)
			} else {
				temp, cnvr = conv(true)
			}
		case 2:
			temp, cnvr = conv(ptr)
		case 3:
			temp, cnvr = conv(t.Fmem[ptr])
		case 4:
			temp, cnvr = conv(t.Cmem[ptr])
		case 5:
			temp, cnvr = conv(t.Bmem[ptr])
		case 6:
			temp, cnvr = conv(string(t.Bmem[ptr]))
		default:
			return "", fmt.Errorf("invalid type : %d", tp)
		}

		if cnvr != nil {
			return "", cnvr
		}
		toadd = toadd + temp + end
		out[num] = toadd
	}
	stro := strings.Join(out, "")
	return stro, nil
}

// get value, [index, type, ptr] by name
func (t *toolbox) Get(name string) (kdbvar, []int) {
	num, cont := t.Name[strings.Replace(name, "/", ".", -1)]
	var out kdbvar
	if !cont {
		return out, nil
	}
	tp := int(t.Tp[num] % 16)
	ptr := t.Ptr[num]

	switch tp {
	case 0:
		out = Set(nil)
	case 1:
		if ptr == 0 {
			out = Set(false)
		} else {
			out = Set(true)
		}
	case 2:
		out = Set(ptr)
	case 3:
		out = Set(t.Fmem[ptr])
	case 4:
		out = Set(t.Cmem[ptr])
	case 5:
		out = Set(t.Bmem[ptr])
	case 6:
		out = Set(string(t.Bmem[ptr]))
	default:
		return out, nil
	}
	return out, []int{num, tp, ptr}
}

// revice data by name
func (t *toolbox) Fix(name string, v interface{}) error {
	num, extv := t.Name[strings.Replace(name, "/", ".", -1)]
	if !extv {
		return fmt.Errorf("KeyError : %s", name)
	}
	end := (t.Tp[num] / 16) * 16

	switch x := interface{}(v).(type) {
	case nil:
		t.Tp[num] = end + 0
		t.Ptr[num] = 0
	case bool:
		t.Tp[num] = end + 1
		if x {
			t.Ptr[num] = 1
		} else {
			t.Ptr[num] = 0
		}
	case int:
		t.Tp[num] = end + 2
		t.Ptr[num] = x
	case float64:
		t.Tp[num] = end + 3
		t.Ptr[num] = len(t.Fmem)
		t.Fmem = append(t.Fmem, x)
	case complex128:
		t.Tp[num] = end + 4
		t.Ptr[num] = len(t.Cmem)
		t.Cmem = append(t.Cmem, x)
	case []byte:
		t.Tp[num] = end + 5
		t.Ptr[num] = len(t.Bmem)
		t.Bmem = append(t.Bmem, x)
	case string:
		t.Tp[num] = end + 6
		t.Ptr[num] = len(t.Bmem)
		t.Bmem = append(t.Bmem, []byte(x))
	default:
		return fmt.Errorf("invalid type : %s", x)
	}
	return nil
}

// convert data -> kformat str
func conv(v interface{}) (string, error) {
	var out string
	switch x := interface{}(v).(type) {
	case nil:
		out = "None"
	case bool:
		if x {
			out = "True"
		} else {
			out = "False"
		}
	case int:
		out = fmt.Sprint(x)
	case float64:
		out = fmt.Sprintf("%f", x)
		for out[len(out)-1] == '0' {
			out = out[0 : len(out)-1]
		}
		if out[len(out)-1] == '.' {
			out = out + "0"
		}
	case complex128:
		st := fmt.Sprint(x)
		out = st[1 : len(st)-1]
	case []byte:
		out = "'" + kio.Bprint(x) + "'"
	case string:
		out = strings.Replace(x, "#", "##", -1)
		out = strings.Replace(out, " ", "#s", -1)
		out = strings.Replace(out, "\n", "#n", -1)
		out = strings.Replace(out, "\"", "#\"", -1)
		out = "\"" + out + "\""
	default:
		return "", fmt.Errorf("invalid type : %s", x)
	}
	return out, nil
}

// add DB by name, var, end
func (t *toolbox) add(name string, v string, end string) {
	nm := strings.Replace(name, "/", ".", -1)
	num := 0
	for nm[num] == '.' {
		num = num + 1
	}
	t.working = append(t.working[0:num], nm[num:])
	nm = strings.Join(t.working, ".")

	var tp byte
	if end == "\n" {
		tp = 0
	} else {
		tp = 16
	}
	tgtv := v

	if tgtv == "None" {
		t.Name[nm] = len(t.Tp)
		t.Tp = append(t.Tp, tp+0)
		t.Ptr = append(t.Ptr, 0)
	} else if tgtv == "True" {
		t.Name[nm] = len(t.Tp)
		t.Tp = append(t.Tp, tp+1)
		t.Ptr = append(t.Ptr, 1)
	} else if tgtv == "False" {
		t.Name[nm] = len(t.Tp)
		t.Tp = append(t.Tp, tp+1)
		t.Ptr = append(t.Ptr, 0)
	} else if tgtv[0] == '"' {
		t.Name[nm] = len(t.Tp)
		t.Tp = append(t.Tp, tp+6)
		t.Ptr = append(t.Ptr, len(t.Bmem))
		t.Bmem = append(t.Bmem, []byte(tgtv[1:len(tgtv)-1]))
	} else if tgtv[0] == '\'' {
		temp, _ := kio.Bread(tgtv[1 : len(tgtv)-1])
		t.Name[nm] = len(t.Tp)
		t.Tp = append(t.Tp, tp+5)
		t.Ptr = append(t.Ptr, len(t.Bmem))
		t.Bmem = append(t.Bmem, temp)
	} else if tgtv[len(tgtv)-1] == 'i' {
		t.Name[nm] = len(t.Tp)
		t.Tp = append(t.Tp, tp+4)
		t.Ptr = append(t.Ptr, len(t.Cmem))
		cv, _ := strconv.ParseComplex(tgtv, 128)
		t.Cmem = append(t.Cmem, cv)
	} else if strings.Contains(tgtv, ".") {
		t.Name[nm] = len(t.Tp)
		t.Tp = append(t.Tp, tp+3)
		t.Ptr = append(t.Ptr, len(t.Fmem))
		fv, _ := strconv.ParseFloat(tgtv, 64)
		t.Fmem = append(t.Fmem, fv)
	} else {
		t.Name[nm] = len(t.Tp)
		t.Tp = append(t.Tp, tp+2)
		iv, _ := strconv.Atoi(tgtv)
		t.Ptr = append(t.Ptr, iv)
	}
}

// import and set by input names & datas & ends (end : \n ;)
func (t *toolbox) Imp(names []string, datas []kdbvar, ends []string) error {
	for i := 0; i < len(ends); i++ {
		name := (names)[i]
		data := (datas)[i]
		var end byte

		num := len(t.Tp)
		t.Name[name] = num
		if ends[i] == "\n" {
			end = 0
		} else {
			end = 16
		}
		t.Tp = append(t.Tp, 0)
		t.Ptr = append(t.Ptr, 0)

		switch data.Dat0 {
		case "None":
			t.Tp[num] = end + 0
			t.Ptr[num] = 0
		case "bool":
			t.Tp[num] = end + 1
			if data.Dat1 {
				t.Ptr[num] = 1
			} else {
				t.Ptr[num] = 0
			}
		case "int":
			t.Tp[num] = end + 2
			t.Ptr[num] = data.Dat2
		case "float":
			t.Tp[num] = end + 3
			t.Ptr[num] = len(t.Fmem)
			t.Fmem = append(t.Fmem, data.Dat3)
		case "complex":
			t.Tp[num] = end + 4
			t.Ptr[num] = len(t.Cmem)
			t.Cmem = append(t.Cmem, data.Dat4)
		case "bytes":
			t.Tp[num] = end + 5
			t.Ptr[num] = len(t.Bmem)
			t.Bmem = append(t.Bmem, data.Dat5)
		case "str":
			t.Tp[num] = end + 6
			t.Ptr[num] = len(t.Bmem)
			t.Bmem = append(t.Bmem, []byte(data.Dat6))
		default:
			return fmt.Errorf("invalid type : %s", data.Dat0)
		}
	}
	return nil
}

// export to precise fullnames, datas, ends (end : \n ;)
func (t *toolbox) Exp() ([]string, []kdbvar, []string) {
	out0 := make([]string, len(t.Tp))
	out1 := make([]kdbvar, len(t.Tp))
	out2 := make([]string, len(t.Tp))
	for name, num := range t.Name {
		out0[num] = name
		tp := t.Tp[num]
		ptr := t.Ptr[num]
		if tp/16 == 0 {
			out2[num] = "\n"
		} else {
			out2[num] = ";"
		}
		tp = tp % 16

		switch tp {
		case 0:
			out1[num] = Set(nil)
		case 1:
			if ptr == 0 {
				out1[num] = Set(false)
			} else {
				out1[num] = Set(true)
			}
		case 2:
			out1[num] = Set(ptr)
		case 3:
			out1[num] = Set(t.Fmem[ptr])
		case 4:
			out1[num] = Set(t.Cmem[ptr])
		case 5:
			out1[num] = Set(t.Bmem[ptr])
		case 6:
			out1[num] = Set(string(t.Bmem[ptr]))
		default:
			return nil, nil, nil
		}
	}
	return out0, out1, out2
}
